### PR TITLE
feature: allow AST to be passed in to IntlProvider, fixes #1335

### LINKED ISSUE #1335: Provide a way pass in already parsed messages by messageformat-parser
**Is your feature request related to a problem? Please describe.**
I need a way to provide precomputed AST instead of string to `IntlProvider`

**Describe the solution you'd like**
Either a new Provider component or overload for `IntlConfig`

```ts
// now
export interface IntlConfig {
  messages: Record<string, string>;
  // other props
}

// proposed
export interface IntlConfigV2 {
  messages: Record<string, string> | Record<string, AST>;
  // other props
}
```

**Describe alternatives you've considered**
Seems that it is not possible today?

**Additional context**
My understanding that the format used internally is  https://github.com/messageformat/messageformat/tree/master/packages/parser.

In our project we have an ability to precompute AST produced by it and save a little bit of time at runtime. Thus the ask to get a first class API support for that :)

Brainstorming potential API imlementations:
1. Accept a new shape for messages
```ts
export interface IntlConfig {
  messages: Record<string, string> | Record<string, AST>;
  // other props
}
```
Cons: maybe hard to disambiguate at runtime?

2. New Provider component
```ts

export const NewIntlProvider : React.FC<NewIntlConfig>  = props => {....} 
export interface NewIntlConfig {
  parsedMessages: Record<string, AST>;
  // other props
}
```
Cons: wider api surface area?

3. Make a new prop `parsedMessages` but require that only one of them would be present
```ts
export interface IntlConfig {
  messages?: Record<string, string>;
  parsedMessages?: Record<string, AST>;
  // other props
}
```

Cons: you can skip both of them and get only a runtime error (assuming ts usage). One way to type it is something like:
```ts
interface A { a: string, b?: never }
interface B { a?: never, b: string }
type Real = A | B
```

### PR DESCRIPTION
**Pull request by longlho**

### CODE REVIEW

File: docs/API.md [hunk 1]
```diff
@@ -394,6 +394,30 @@ The message formatting APIs go the extra mile to provide fallbacks for the commo
 
 Above, "source" refers to using the template as is, without any substitutions made.
 
+#### Advanced Usage
```
Comment (redonkulus) [inline docs/API.md]: Add link to table of contents above

Comment (longlho) [reply docs/API.md]: I automated it

File: scripts/toc.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+const {sync: globSync} = require('glob');
+const {resolve} = require('path');
+const toc = require('markdown-toc');
+const {outputFileSync} = require('fs-extra');
+const {readFileSync} = require('fs');
+globSync(resolve(__dirname, '../docs/*.md')).forEach(fn => {
```
Comment (redonkulus) [inline scripts/toc.js]: if this is run again will output 2 TOC's or is it smart enough to update the existing one?

File: package.json [hunk 3]
```diff
@@ -132,6 +134,7 @@
     "watchify": "^3.7.0"
   },
   "scripts": {
+    "toc": "node scripts/toc",
```
Comment (redonkulus) [inline package.json]: alphabetize

File: docs/Upgrade-Guide.md [hunk 4]
```diff
@@ -1,14 +1,37 @@
+## Table of Contents
```
Comment (redonkulus) [inline docs/Upgrade-Guide.md]: I still like the title to be above the TOC, so the user knows what page they are on

File: scripts/toc.js (hunk 2 above)
Comment (longlho) [reply scripts/toc.js]: nope it checks for the `<!-- toc -->` thing

**Pull request merged**